src/jetstream_push_subscription.rs

```diff
@@ -0,0 +1,544 @@
+        self.0
+            .context
+            .connection
+            .0
+            .client
+            .flush(DEFAULT_FLUSH_TIMEOUT)?;
+
+        self.0
+            .context
+            .connection
+            .0
+            .client
+            .unsubscribe(self.0.sid.load(Ordering::Relaxed))?;
+
+        // Discard all queued messages.
+        while self.0.messages.try_recv().is_ok() {}
+
+        // Delete the consumer, if we own it.
+        if self.0.consumer_ownership == ConsumerOwnership::Yes {
+            self.0
+                .context
+                .delete_consumer(&self.0.stream, &self.0.consumer)?;
+        }
+
+        Ok(())
+    }
+
+    /// Close a subscription. Same as `unsubscribe`
+    ///
+    /// Use `drain` instead if you want any pending messages
+    /// to be processed by a handler, if one is configured.
+    ///
+    /// # Example
+    /// ```
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// let sub = nc.subscribe("foo")?;
+    /// sub.close()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn close(self) -> io::Result<()> {
+        self.unsubscribe()
+    }
+
+    /// Send an unsubscription then flush the connection,
+    /// allowing any unprocessed messages to be handled
+    /// by a handler function if one is configured.
+    ///
+    /// After the flush returns, we know that a round-trip
+    /// to the server has happened after it received our
+    /// unsubscription, so we shut down the subscriber
+    /// afterwards.
+    ///
+    /// A similar method exists on the `Connection` struct
+    /// which will drain all subscriptions for the NATS
+    /// client, and transition the entire system into
+    /// the closed state afterward.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use std::sync::{Arc, atomic::{AtomicBool, Ordering::SeqCst}};
+    /// # use std::thread;
+    /// # use std::time::Duration;
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    ///
+    /// let mut sub = nc.subscribe("test.drain")?;
+    ///
+    /// nc.publish("test.drain", "message")?;
+    /// sub.drain()?;
+    ///
+    /// let mut received = false;
+    /// for _ in sub {
+    ///     received = true;
+    /// }
+    ///
+    /// assert!(received);
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn drain(&self) -> io::Result<()> {
+        self.0
+            .context
+            .connection
+            .0
+            .client
+            .flush(DEFAULT_FLUSH_TIMEOUT)?;
+
+        self.0
+            .context
+            .connection
+            .0
+            .client
+            .unsubscribe(self.0.sid.load(Ordering::Relaxed))?;
+
+        // Delete the consumer, if we own it.
+        if self.0.consumer_ownership == ConsumerOwnership::Yes {
+            self.0
+                .context
+                .delete_consumer(&self.0.stream, &self.0.consumer)?;
```

Comment:
Double unsubscribing is a failure, probably should be a no-op.

Reply:
In Go client double unsubscribe does not error:
```Go
	// Already unsubscribed
	if s == nil {
		return nil
	}
```
So I will follow the same approach.